### DISCUSSION TITLE
AsMutable extension method

### DISCUSSION THREAD

**Discussion by emphasis87** [dotnet/runtime#84072] [Ideas] [answered]
Would it be possible to have a shorthand extension method to convert IReadOnly types (set, dictionary, list) back to mutable ones?

**Comment by huoyaoyuan** [comment, marked as answer]
No, because same data types are designed without mutable support in order to minimize copying, especially the `Immutable*` ones.

**Reply by emphasis87** [reply]
That is really not the point. The point is to not have to write new Dictionary<string, string>(readOnlyDictionary), or be able to use readOnlyDictionary?.AsMutable() ?? new Dictionary<string, string>() and so on.

**Reply by emphasis87** [reply]
Currently if you have readOnlyDictionary variable, you have to write cumbersome code like this:
`var d = readOnlyDictionary is { } ? new Dictionary<string,string>(readOnlyDictionary) : new Dictionary<string, string>();`

**Reply by emphasis87** [reply]
It seemed to me that AsMutable could be a nice counterpart to AsReadOnly extension method, which are clear in allocating a new object.

**Reply by huoyaoyuan** [reply]
> `readOnlyDictionary?.AsMutable() ?? new Dictionary<string, string>()`

Such code is really confusing. You get a mutable collection with the same content of your given readonly collection, but whether the changes is observable with the original readonly collection reference is not determined.

If you have the requirement to use readonly collections and have mutated versions of them, does `ImmutableDictionary` or `FrozenDictionary` meet your requirement?

**Reply by huoyaoyuan** [reply]
> It seemed to me that AsMutable could be a nice counterpart to AsReadOnly extension method, which are clear in allocating a new object.

`AsReadOnly` has a clear contract that the input and output reference will view the same content and changes. Such constraint can't be satisfied for `AsMutable`.

**Reply by emphasis87** [reply]
Ah, I see. I mistakenly thought that `AsReadOnly<TKey,TValue>(IDictionary<TKey,TValue>)` returns a new collection, but it is only a wrapper. My use case is that I have a public dictionary that should not be edited once populated, and then I need to create another dictionary starting with the same values, then extending it and then again setting it to a public property as not editable.

**Reply by emphasis87** [reply]
At first I tried to solve it by Enumerable.ToDictionary, but that involved unnecessary lambdas. So, anyway, I think I will mark it as answered, for now. Unless ReadOnlyDictionary.ToDictionary extension method becomes a possibility?

**Reply by huoyaoyuan** [reply]
> At first I tried to solve it by Enumerable.ToDictionary, but that involved unnecessary lambdas.

See #65925.
